File: src/day1.cpp
#include <iostream>
#include <fstream>

int main(int argc, char **argv)
{
	if (argc != 2) {
		std::cerr << "Incorrect number of arguments provided\n";
		return 1;
	}
	std::fstream input(argv[1]);
	if (!input) {
		std::cerr << "Could not open input file\n";
		return 1;
	}

	int floor = 0;
	int count = 0;
	int basement_pos = 0;
	char c;
	while (input >> c) {
		count++;
		if (c == '(') {
			floor++;
		} else if (c == ')') {
			floor--;
		}
		if (basement_pos == 0 && floor < 0) {
			basement_pos = count;
		}
	}

	std::cout << "Santa is on floor: " << floor << "\n";
	std::cout << "Santa entered the basement on move: " << basement_pos << "\n";
	return 0;
}

File: 2019/src/bin/day1.rs
use std::env;
use std::fs::File;
use std::io;
use std::io::{BufRead, BufReader};

/* Shiny iterator function */
fn fuel_reqs(mass: i32) -> impl std::iter::Iterator<Item = i32> {
    let mut m = mass;
    std::iter::from_fn(move || {
        m = m / 3 - 2;
        if m > 0 {
            Some(m)
        } else {
            None
        }
    })
}

fn main() -> io::Result<()> {
    if env::args().len() != 2 {
        panic!("Incorrect number of arguments provided");
    }
    let modules: Vec<_> =
        BufReader::new(File::open(env::args().nth(1).unwrap()).expect("Could not open input file"))
            .lines()
            .map(|l| l.unwrap().parse().unwrap())
            .collect();

    let fuel_sum: i32 = modules.iter().map(|n| fuel_reqs(*n).next().unwrap()).sum();
    let rec_fuel_sum: i32 = modules.iter().map(|n| fuel_reqs(*n).sum::<i32>()).sum();
    println!("Fuel requirements: {}", fuel_sum);
    println!("Fuel requirements (inc fuel): {}", rec_fuel_sum);
    Ok(())
}
